dev/Dockerfile

```diff
@@ -29,26 +29,40 @@ ENV PATH "${DENO_INSTALL}/bin:${PATH}"
 
 COPY --from=deno-bin /usr/bin/deno /bin/deno
 
-RUN apt-get update \
-    && apt-get install -y --no-install-recommends \
-    # base
-    git \
-    curl \
-    python3 \
-    make \
-    cmake \
-    # grpc
-    protobuf-compiler \
-    libprotobuf-dev \
-    # openssl
-    pkg-config \
-    libssl-dev \
-    # wasmedge
-    libclang-dev \
-    && rm -rf /var/lib/apt/lists/* \
-    && curl -fsS https://raw.githubusercontent.com/WasmEdge/WasmEdge/master/utils/install.sh -o wasmedge.sh \
-    && bash wasmedge.sh -v $WASMEDGE_VERSION \
-    && rm wasmedge.sh
+RUN set -eux; \
+    export DEBIAN_FRONTEND=noninteractive; \
+    apt update; \
+    apt install --yes --no-install-recommends \
+        # base \
+        git \
+        curl \
+        make \
+        # openssl deps \
+        pkg-config \
+        libssl-dev \
+        # wasmedge deps \
+        libclang-dev \
+        # asdf deps
+        xz-utils \
+        unzip \
+        # poetry deps \
+        python3 \
+        python3-setuptools \
+        python3-venv \
```

Comment:
Actually, even python doesn't appear to be used by the Dockerfile and we should be able to yeet it. Regarding excluding ports, you might have seen the environment variable tests I use in `ghjk.ts` to exclude some env vars. We can definitely formalize this as part of the _ports module_  with the flags like the one you describe weather it's exclusion of user specified ports or maybe, port sets.

To bikeshed some on the syntax there though, the `envs` array as you wrote it there won't exactly work with the current approach since `deno` and `node` are just port specific functions that simply add an `InstallConfig` to the global `installs` list.

Maybe syntax like this might help if we ditch the global approach:

```ts
export ports = [
    node({ version: "galdonium" }),
    rust({ targets: ["wasm-32"] }),
];

export tasks = [
    task(
        "install-gate", 
        [deno(), rust()] ,
        async () => await $`cargo ...`, 
     ),
];
```

But I think I've come to prefer how the current global ux looks now. 
Here are some concerns though:

- To keep it flexible, any `ghjk.ts` will be acceptable as long as they export an object [this](https://github.com/metatypedev/ghjk/blob/46fa1895dff4e4e2bb18ed92b1c6bb68c3e9dbf9/mod.ts#L56) so we should be able to experiment with the syntax all we like.
    - Currently, the default exported functions ([example](https://github.com/metatypedev/ghjk/blob/46fa1895dff4e4e2bb18ed92b1c6bb68c3e9dbf9/ports/node.ts#L39)) are not hard required by the impl. Making it a hard-convention is attractive but maybe someone might want to write some port that adds 15 installs and wants to provide a builder interface for that instead. 
    - A use-case I've been thinking about is allowing the ports to be consumable without making use of ghjk environments. I.e. some script that takes a URL to a port and installs it globally in the XDG dirs or something. 
        - Instead of `curl sh.pnpm.com | sh`, think `deno run ghjk.dev/standalone ghjk.pnpm.com --version=12 --port-custom-conf=abc`.  
        - Those flags will be turned into a json `InstallConfig` of the shape `{ "version": "12", "port-custom-conf": "abc" }`. 
        - Deno will prompt the user for every permission the port needs
        - This would be complicated if those default export functions or whatever we replace them with did anything more interesting than add an `InstallConfig` into the global array. 


In fact, I think we might be able to hide away the global impl detail from the ports impls entirely:

```ts
// my-port.ts

// note: this returns an array to allow multiple installs
export default function conf(args: CustomInstallArgs): InstallConfig[] {
    return { version: args.version, ...args, portMainfest: manifest };
}

// ghjk.ts
import myPort from ".....my-port.ts"

// use the sugar function from the port
ports(
       myPort({version: "abc"})
);
// which is pretty much doing this
ports({ version: "asdf", ty: "deno", moduleSpecifier: "http://abc.abc" });

// the `ports` function manipulates the global var instead
// it also dedupes the all the `PortManifests` coming
// from the all `InstallConf`

// i suppose this strays from the current purely global syntax
// but it doesn't look too bad.
ports(
    myPort({version: "abc"}),
    myPort({version: "abc2"}),
    myPort({version: "abc3"}),
);

// people who want it to be more conf like won't like it
// but it's less magical for js trained eyes.
ports(   
    myPort({version: "abc"}),
    myPort({version: "abc2"}),
    myPort({version: "abc3"}),
)
 // apply overrides to the entire InstallConfig set from above 
 .overrides({ set: "ci", env: {custom: "env"} });

```

And maybe something similar for tasks:
```ts

tasks({
  name: "my-task-1",
  ports: [myPort({version: "123" })], // note[!]
  do: async () => {
    await $`cargo ...`;
    await fetch("http://...");
  },
}); // we can add more tasks in one call too

// we can also let users import special task helpers similar to the port default exports
// but this is all an implementation detail

// my-task-helper.ts
export default function helper(args: CustomTaskArgs): TaskConf {
    return { /* ... */ };
}
```

That's only a quick sketch, the only important change being that the port impls are no longer aware of the global variables.

Reply:
Let's discuss this on our next call, as there is much to say :) 

I have the feeling the design could change drastically when we support task, because most of the teams the ports will only by used within a given task and not outside.

> ports to be consumable without making use of ghjk environments
Interesting, let's keep this in mind (yet I would argue this would make sense only when envs are mature)

> port impls are no longer aware of the global variables
I am not sure to follow, can you dig into that catch?

"Frontend"-wise, I am not a huge fan of having overrides or any non-obvious behaviour. Taking again the Metatype example with the 2 versions of rust, I would love to be able to manipulate the object directly (the graph ought to be explicit) - something like:
```ts
myCustomRust = rust({...})
nightlyRust = rust({...})

buildWasm = task({
  ..,
  envs: [nightlyRust]
})

buildTypegate = task({
  ..,
  deps: [buildWasm],
  envs: [myCustomRust]
})
```

Reply:
Sure, we'll discuss in the call. Good point about how most ports will only ever be used with tasks.

> port impls ... global variables
> not sure to follow

Currently, ports are expected to add themselves to global arrays of [ports](https://github.com/metatypedev/ghjk/blob/46fa1895dff4e4e2bb18ed92b1c6bb68c3e9dbf9/ports/protoc.ts#L24) and [installs](https://github.com/metatypedev/ghjk/blob/46fa1895dff4e4e2bb18ed92b1c6bb68c3e9dbf9/ports/protoc.ts#L27). I just want to remove that for some of the reasons outlined. A very good reason is what you mention, if most ports are used for tasks, the global variables even more undesirable.

Instead of adding items to the global variable, they return simple `InstallConf` objects which brings me to other point....

> overrides

What the override function does in that sketch there is only just modify the `InstallConf` objects created by the ports. I like the snippet you have there there and to build on it:

```ts
// this are InstallConf objects which aren't classes
// but pure json
const myCustomRust = rust({...})
const nightlyRust = rust({...})

// specify that we want this in our env as well
// note: i renamed ports from above to `install`
install(
    nightlyRust,
    myCustomRust,
    node(),
)

// that override function is just a shorthand for things like these
install(
     rust({
        // what env vars are avail to a Port install have to be
        /// specified up front (either by the port default fn or the user)
        // here, we modify env vars specified by `rust({...}` 
        env: { CARGO_TARGET_DIR: "target/wasm" },
        // we include it only in the ci set to allow features like
        // `ghjk ports sync --set ci`
        // sets are like docker compose profiles
        set: ["ci"],
      }),
     // InstallConfs are just pure jsons so above is identical to
    {
        ...myCustomRust,
        env: { CARGO_TARGET_DIR: "target/wasm", ...myCustomRust.env },
        set: ["ci"],
    },
)
    // this is a shorthand for modifying the installs from above at once
    .override({ set: ["ci"] })


```

Reply:
Alternative syntax:

```ts
install(
    set({ profiles: ["ci"], env:{} },
        node(),
        rust(),
    )
)
```

Reply:
Another pickle in ghjk we have to figure out is runtime install dependencies as opposed to build time port dependencies. `jco` required `node` to function at runtime. Right now, I'm simply letting the `jco` port also add an install for `node` but this can get messy for the following reasons:

- The default export fns like`node`, `rust` and `jco` there have to return `InstallConf[]` now to supprt this. I.e. `install` takes `InstallConf[][]`.
- if multiple ports rely at runtime on the same install, we'll have clashes. Either one of the port authors will have to opt-in and provide some config var to disable their install or we'll have to clean up those `InstallConf[]` from ambove.
- if we want to customize the `node` install, we'll have to go through `jco` or instruct `jco` not to add `node` and add it ourselves.
- `node` will now be avail in our environment even though if we don't want it. If there's a global `node` install we prefer, it'll shadow it

One way we can get around this is if the `jco` plugin replaced it's shebangs with something like `#/path/to/runtime/node` instead of `#/usr/bin/env node`. But that can be a bit involved and not everything is as simple as shebangs (think LD_LIBRARY runtime deps). It's also explicit thing port authors will have to be aware and opt-in for. 

Another approach is to replace all the symlink based shims with an sh scripts based shims when we encounter these kinds of cases. This way, `jco` is no longer a symlink to `jco.js` but a shellscript like the following:
```sh 
#! /bin/sh
PATH=/runtime/dep/bin/shims:$PATH 
LD_LIBRARY_PATH=/runtime/dep/lib/shims:$LD_LIBRARY_PATH
# we can even add custom runtime env vars here
/path/to/jco.js $argv
``` 

Another way fwd is to keep the current `InstallConf[]` approach and users who don't want to shadow/pollute their env can wrap these polluting deps in a task:

```ts
tasks(
    name:  "jco-sandbox",
    deps: [jco({ /* .. */ })],
    do: ({ argv }) => $`jco ${argv}` // we can add custom env vars here
) 
// we might be able to supply helpers to easily do this
tasks(sandboxInstall(jco(), 'jco'))
```

Reply:
:+1: for the config, then act when the graph is built. Why not call `InstallConf` `Port`?

A port shall specify its dependencies with a "kind" and a version/range that we can fuzzy match. If you install jco, I expect to have a failure if the node dependency is not specified explicitly `jco({deps: [node])}` (type hints will help the users). We can also allow default dependencies if the version/range is wide enough (`*` ok, `>15`, else not) and match them against other node instance installed in the same project. If they do not fit or if the matching cannot be done, we throw and ask the user to specify. It will be likely that there is a set of original port required by most of the project (e.g. most of the ambient ones) and I would pre-package them as a one-liner:
```ts
const gen = ghjk.default_ports()
gen.git // git({version: "*"})
```

I am not yet sure about the shell script shim approach because it may raise more questions/challenges than having a single env setup that works everywhere: a shim A could use B setting some specific envs that would later be overridden by shim B, and we lose the easy "ln" explanation.

Reply:
This is being addressed in [#14](https://github.com/metatypedev/ghjk/pull/14) and should reflect in this repo after that's done.